contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,417 @@
+  VolumeHandle string `json:"volumeHandle" protobuf:"bytes,2,opt,name=volumeHandle"`
+
+  // Optional: MountSecretRef is a reference to the secret object containing
+  // sensitive information to pass to the CSI driver during NodePublish. 
+  // This may be empty if no secret is required. If the secret object contains
+  // more than one secret, all secrets are passed.
+  // This is a local object reference so only secrets within the same
+  // namespace as the pod or PVC referencing this volume can be referenced.
+  // +optional
+  MountSecretRef *LocalObjectReference `json:"secretRef,omitempty" protobuf:"bytes,3,opt,name=secretRef"`
+
+  // Optional: AttachSecretRef is a reference to the secret object containing
+  // sensitive information to pass to the CSI driver during ControllerPublish. 
+  // This may be empty if no secret is required. If the secret object contains
+  // more than one secret, all secrets are passed.
+  // +optional
+  AttachSecretRef *ObjectReference `json:"secretRef,omitempty" protobuf:"bytes,3,opt,name=secretRef"`
+
+  // Optional: The value to pass to ControllerPublishVolumeRequest.
+  // Defaults to false (read/write).
+  // +optional
+  ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,4,opt,name=readOnly"`
+}
+```
+
+#### Internal Interfaces
+
+The in-tree CSI volume plugin will implement the following internal Kubernetes volume interfaces:
+
+1. `VolumePlugin`
+    * Mounting/Unmounting of a volume to a specific path.
+2. `AttachableVolumePlugin`
+    * Attach/detach of a volume to a given node.
+
+Notably, `ProvisionableVolumePlugin` and `DeletableVolumePlugin` are not implemented because provisioning and deleting for CSI volumes is handled by an external provisioner. 
+
+#### Mount and Unmount
+
+The in-tree volume plugin’s mount and unmount methods will trigger the `NodePublishVolume` and `NodeUnpublishVolume` CSI calls via Unix Domain Socket. Kubernetes will generate a unique `target_path` (unique per pod per volume) to pass via `NodePublishVolume` for the CSI plugin to mount the volume. Upon successful completion of the `NodeUnpublishVolume` call (once volume unmount has been verified), Kubernetes will delete the directory.
+
+The Kubernetes volume sub-system does not currently support block volumes (only file), so for alpha, the Kubernetes CSI volume plugin will only support file.
+
+#### Attaching and Detaching
+
+The attach/detach controller,running as part of the kube-controller-manager binary on the master, decides when a CSI volume must be attached or detached from a particular node.
+
+When the controller decides to attach a CSI volume, it will call the in-tree CSI volume plugin’s attach method. The in-tree CSI volume plugin’s attach method will do the following:
+
+1. Create a new `VolumeAttachment` object (defined in the “Communication Channels” section) to attach the volume.
+    * The name of the of the `VolumeAttachment` object will be the `<UniqueVolumeName>-<NodeId>`.
```

Comment:
Do we have any other naming patterns already used by K8S components?

Reply:
None that would apply here. Usually it's user who chooses object name. Kubernetes adds random suffix or sequence number to Pod.Name created from a ReplicaSet or StatefulSet. We need the name of VolumeAttachment to be fixed to catch collisions when something wants to attach the same volume to the same node twice - VolumeAttachment objects will have the same name and thus only one can be successfully created.

Reply:
Are UniqueVolumeName or NodeId able to contain a `-` character?

Reply:
It doesn't matter if they do. The `-` character is not used as a delimiter. The generated name is simply used as a deterministic way of ensuring that attach operations are not triggered for the same volume more then once.

Reply:
It could matter if one unique volume name could be a prefix of another. 

volumename foo and namespace bar-baz
volumename foo-bar and namespace baz

Those would conflict and mean that one volume attach would not be able to be satisfied